ref(workflow_engine): Adding helpers for logs / metrics

saponifi3d opened the pull request:
# Description
Started to refactor `process_workflows` metrics / logging to be improved.

This lead me to not wanting to share data by passing it through methods just for metrics / logs. Paused the refactor to then build out `WorkflowContext` which can share information across app during the life cycle of `process_workflows` (for now, i'll add it to `process_detectors` as well in the future. those contexts will drop out of scope and be cleaned up automatically in each process)

Also created a `metrics_incr` method to illustrate the use of `WorkflowContext` and consumed that in `action.trigger` as a way to clean-up process_workflows.

---

I thought it might be useful to see all of the usage together, but let me know if it would be easier to split this into multiple PRs for review.

saponifi3d committed:
Create a utils file for the metrics_incr method that prefixes everything
for workflow engine.

saponifi3d committed:
make a utils directory

saponifi3d committed:
start adoption of the new util that will prefix everythign with  and have a centralized metric method

saponifi3d committed:
WIP

saponifi3d committed:
start to use the workflow context / use it in a few places and see what breaks

saponifi3d committed:
simplify the metrics w/o the caller in the stack. dont need to go so ham

saponifi3d committed:
clean up the action.trigger

saponifi3d committed:
encapsulate the context variables within the class

saponifi3d committed:
change class test name and the file name to try and fix tests

saponifi3d committed:
cleanup some test stuff

saponifi3d committed:
fix mypy types

saponifi3d committed:
update the processor tests with the changes to metrics / logs so far

saponifi3d committed:
ensure that we have a clear context for each test.

ceorourke commented on src/sentry/workflow_engine/processors/workflow.py:
```diff
@@ -112,6 +116,27 @@ def evaluate_workflow_triggers(
+        tags={"detector_type": detector.type},
+    )
+
+    environment = get_environment_by_event(event_data)
+    if environment is None:
```
Why do we need to refetch this when we get the environment on lines 101-103?

ceorourke commented on src/sentry/workflow_engine/processors/workflow.py:
```diff
@@ -112,6 +116,27 @@ def evaluate_workflow_triggers(
+
+    environment = get_environment_by_event(event_data)
+    if environment is None:
+        return set()
```
What if the workflow just doesn't have an environment set? That seems like it'd be ok and we don't want to exit early.

ceorourke replied:
Ah I see, yeah `Environment.DoesNotExist` sounds good

ceorourke commented on src/sentry/workflow_engine/utils/metrics.py:
```diff
@@ -0,0 +1,51 @@
+from typing import Any
```
What's the purpose of the empty `logs.py` file?

saponifi3d replied:
that's where i'm going to add the shared workflow logger stuff - whoops tho, i accidentally made it then changed branches. deleting ❌

ceorourke reviewed (approved):
mostly lgtm, just a few comments

saponifi3d committed:
rather than return none, re-raise the error after logging. this makes the control flow a lot clearer

saponifi3d committed:
:sweat-smile: whoops, preview file.

saponifi3d committed:
move the context vars out of the class

saponifi3d committed:
whoops, fix the return type

saponifi3d committed:
whoops, missed this bit of logic

saponifi3d committed:
another one

saponifi3d committed:
we can just ues the log from get_detector_by_event

saponifi3d committed:
add a uuid for the context, this will allow us to uniquely trace each log

saponifi3d committed:
it works, but the keys aren't dynamic...

saponifi3d commented on src/sentry/workflow_engine/utils/metrics.py:
```diff
@@ -0,0 +1,51 @@
+    return f"{METRIC_PREFIX}.{metric_name}"
+
+
+def metrics_incr(
```
😅 sorry, tried to explain this a little in the PR description. I was starting down a refactor and then popped up to this. rather than making this PR more confusing, i opt'd to open the pr with the implementation and 1 example use. my plan is to open up an adoption pr after to try and make this a bit more readable for reviews.

saponifi3d committed:
now with improved method names and doc strings

saponifi3d committed:
well, that probably solved a future issue. this change will use the ctx_tags and then override with anything passed in

saponifi3d commented on src/sentry/workflow_engine/utils/metrics.py:
```diff
@@ -0,0 +1,51 @@
+    value (int): The number to increment by
+    tags: MetricTags | None: Optional tags to add to the metric.
+    """
+    ctx = WorkflowContext.get()
```
🤔 i generally like the idea, but i think the dict would be fairly different between logs and metrics since we need to limit in metrics to avoid high cardinality.

since we can't share the `as_tags` method, it feels like it'd be fairly similar to provide that method or decorate a centralized log.

-- also want to make sure, is that what you were thinking the use would be for?

saponifi3d commented on src/sentry/workflow_engine/utils/workflow_context.py:
```diff
@@ -0,0 +1,68 @@
+    from sentry.models.organization import Organization
+    from sentry.workflow_engine.models import Detector
+
+_detector: ContextVar[Detector | None] = ContextVar("detector", default=None)
```
... i really wish i had made the dataclass first / seen this! 🤣 -- i like that idea. updating!

saponifi3d committed:
Update WorkflowContext to interact with a frozen data-set. gets all the advantages from before but simpler

saponifi3d commented on src/sentry/workflow_engine/utils/workflow_context.py:
```diff
@@ -0,0 +1,68 @@
+        return value is not None or variable.get() is not None and value is None
+
+    @classmethod
+    def set(
```
did a bit of a refactor here -- `_check_value` is gone, curious your thoughts with the frozen dataclass suggestion 🌮

saponifi3d committed:
Fix the test / mypy error

kcons commented on src/sentry/workflow_engine/utils/workflow_context.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class WorkflowContextData:
```
Maybe here, or maybe at `WorkflowContext`, it seems important to give some usage guidance.
When should you rely on context values, vs when should you expect them to be passed in?
Should passed values be assumed consistent with `WorkflowContext` values? Is it a bug if they're different?
Do we have an obligation to keep `WorkflowContext` values up-to-date? eg, if I have a loop over various workflows, should I be setting `Environment` for each one?

saponifi3d committed:
Address PR Feedback:
- Add a comment block to describe usage of WorkflowContext.
-- This got me thinking more about stuff and refactored a little for
simplicity as well.

saponifi3d committed:
whoops, over zealous deletion, but this shuold be after we execute the action rather than before

saponifi3d committed:
don't have an id anymore

saponifi3d committed:
got ahead of myself witht his change to delayed workflows

saponifi3d committed:
fix the processor tests to include the shared metric info

saponifi3d committed:
fix tests in ci

saponifi3d committed:
set the context for a detector earlier, that way we can use it if there's a problem with the environment

saponifi3d committed:
fix last test

saponifi3d committed:
Merge remote-tracking branch 'origin/master' into jcallender/aci/improved-metrics

saponifi3d committed:
fix test after merge conflict
